README.md
# IPK-p1

## About app
- `ipkcpc` application is client for communication with dedicated calculator server
- Application is meant for UNIX systems, but can be run on Windows (more in point `Windows`)

## Message length limits
- UDP
    - Send: 254 Characters
    - Recieve: 253 Characters
- TCP
    - Send: 255 Characters (Including LF == '\n')
    - Recieve: 255 Characters (Including LF == '\n')

## Translation
- in parent directory
    ```
    $ make
    $ make run
    $ make clean
    ```
- And special make for printing DEBUG informations
    ```
    $ make debug
    ```

## Execution
```bash
$ ./ipkcpc [parameters] [options] ...
```
#### parameters
- `-h` - Host name (`IPv4` address) in from `X.X.X.X`, where `X` is in range <0,256>. 
[`0.0.0.0` by default]
- `-p` - Port for connection to the server, number in range <0,65535>.
[`2023` by default]
- `-m` - Which type of connection will be used. TCP/UDP.
[`TCP` by default]
#### stdin
- After execution, messages for the server will be taken from `STDIN`, sent by pressing `ENTER`

#### expected input
- TCP
    - First message for the server has to be `hello`
    - For calculating is message `solve`
    - End communication by `bye`
- UDP
    - There is no `hello`, `solve` or `bye`
    - Calculation is provided only by typing `expr` and pressing `ENTER`
- Syntax:
    ```
    hello = "HELLO" LF
    solve = "SOLVE" SP query LF
    result = "RESULT" SP 1*DIGIT LF
    bye = "BYE" LF
    ```
    ```
    operator = "+" / "-" / "*" / "/"
    expr = "(" operator 2*(SP expr) ")" / 1*DIGIT
    query = "(" operator 2*(SP expr) ")"
    ```
   

## Exit codes
- 0 - Program run without problem
- 1 - Wrong parameters while executing program 
- 2 - Creating socket failure
- 3 - Sending packet failure
- 4 - TCP connection could not be created
- 5 - Packet recieving failure
- 6 - Recieved OPCODE from server was not in range {0,1}
- 7 - Recieved status code was not in range {0,1}
- 8 - WSA creation failure (only for `WINDOWS OS`)
- 99 - Internal program failure (malloc, calloc, etc.)

## Details of communication
#### UDP
- `REQUEST` is send as bytes
    ```
            0                   1                   2                   3
    bites:  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
            +---------------+---------------+-------------------------------+
            |     Opcode    |Payload Length |          Payload Data         |
            |      (8)      |      (8)      |                               |
            +---------------+---------------+ - - - - - - - - - - - - - - - +
            :                     Payload Data continued ...                :
            + - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +
            |                     Payload Data continued ...                |
            +---------------------------------------------------------------+
    ```
- `RESPONSE` is in bytes
    ```
            0                   1                   2                   3
    bites:  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
            +---------------+---------------+---------------+---------------+
            |     Opcode    |  Status Code  |Payload Length | Payload Data  |
            |      (8)      |      (8)      |      (8)      |               |
            +---------------+---------------+---------------+ - - - - - - - +
            :                     Payload Data continued ...                :
            + - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +
            |                     Payload Data continued ...                |
            +---------------------------------------------------------------+
    ```
- `OPCODES` (value of byte)
    ```
    OPCODE       |   VALUE
    ---------------------------
    REQUEST      |     0
    RESPONSE     |     1
    ```

- `STATUS CODES` (value of byte)
    ```
    STATUS CODE  |   VALUE
    ---------------------------
         OK      |     0
        ERROR    |     1
    ```

#### TCP
- TCP is send as text (message format via `EXECUTION`)

## Testing
- In repository in folder `../tests/` are testing scripts and files with input and expected output
- Scripts:
    - `create-test.sh` 
        - running this script will help you to create file in exact folder, but tests itself have to by written by your own hand
        - will create files
            - `"FILE_NAME".in` for your selected input
            - `"FILE_NAME".out` in this file write expected output
            - `"FILE_NAME".errcode` for expected error code
            - `"FILE_NAME".arg` for arguments for execution `ipkcpc` [DONT CHANGE]
    - `remove-test.sh`
        - remove all files (`.in`, `.out`, `.errcode`,`.arg`) with select name
    - `run-tests.sh`
        - will rune all tests from folders in folder `../tests/`
        - operators:
            - `-m` will select which tests will be tested [tcp/udp]
            [`tcp` by default]
            - `-h` IPv4 address of server on which will be tested the tests
            [`0.0.0.0` by default]
            - `-p` on which port will be communication with server
            [`2023` by default]
        - Results of tests will be print on `stdin`
        - example of exectution
            ```bash
            $ ./run-tests.sh -m tcp -h 127.0.0.1 -p 2000
            ```
        
## Windows
- Application can run on windows, too.
- Normal `make` wont work, need to be executed by:
```bash
$ make all
```
- In IDE CLion from JetBrains you are able to execute it easilly or use this to translate:
```
$ gcc -g -std=c11 -Wall -Wextra -Werror -pedantic ipkcpc.c -o ipkcpc -lwsock32 -lws2_32
```
- And support of debug output is available as
```
$ make debug-win
```
- Execution of program on Windows is
```
$ ./ipkcpc.exe [parameters] [options]
```

## Reference links
- https://git.fit.vutbr.cz/NESFIT/IPK-Projekty/src/branch/master/Project%201

## Source
- Program was inspired from NESFIT git:
    - https://git.fit.vutbr.cz/NESFIT/IPK-Projekty/src/branch/master/Stubs
- And from `educative.io`
    - https://www.educative.io/answers/how-to-implement-tcp-sockets-in-c
    - https://www.educative.io/answers/how-to-implement-udp-sockets-in-c

## LICENSE
- This program is under License GPLv3 (<https://fsf.org/>)
- or in repository in file LICENSE

CHANGELOG.md
# CHANGELOG

## Initial commit
- Initial commit was created after few commits on my personal github, so inital commit is a bit bigger
- `ADD:` Testing right shapes of switches and arguments

## First iteration
- `ADD:` TCP connection creating and sending messages
- `ADD:` UDP connection creating and sending messages
- `FIX:` lot of commits were used to repair bugs from TCP and UDP connection

## Second iteration
- `ADD:` Basic testing of source codes, testing scripts and files for tests
- `FIX:` UDP and TCP message sending repairing bugs

## Third iteration
- `ADD:` Windows support

## Fourth iteration
- `ADD:` Links for source codes that this projects was inspired from
- `ADD:` Licensing

## Fifth iteration
- `ADD:` CHANGELOG.md
- `ADD:` timeout on UDP communication

## Limitations of program
- As it is described in documenatation (`README.md`), there is a limitation for message input and output in length
- UDP
    - Send: 254 Characters
    - Recieve: 253 Characters
- TCP
    - Send: 255 Characters (Including LF == '\n')
    - Recieve: 255 Characters (Including LF == '\n')

tests/run-tests.sh
######################################
# 
# Modul:  run-tests.sh
#
# Author: Michal Zapletal
# Email:  xzaple41@stud.fit.vutbr.cz
#
######################################

#!/bin/bash

pushd ..

while getopts p:h:m: flag
do
    case "${flag}" in
        h) ip=${OPTARG};;
        p) port=${OPTARG};;
        m) mode=${OPTARG};;
    esac
done

if ! [ "$ip" = "" ]; then
    arg=$arg"-h "$ip
fi
if ! [ "$port" = "" ]; then
    arg=$arg" -p "$port
fi
if [ "$mode" = "" ]; then
    mode="tcp"
fi
echo "[START] Tests from \"$mode\" folder"

test_ok=0
test_not_ok=0

DIR=`ls ./tests/$mode/*`
#echo $DIR
for file in $DIR;
do
    files=$(exec ls)
    case "$files" in
        *output.txt*) rm output.txt;
    esac
    
    regex=".*\.in"
    if ! [[ $file =~ $regex ]]; then
        continue
    fi
    IFS='/'
    read -ra strarr <<< $file
    IFS='.'
    read -ra strarr <<< ${strarr[3]}
    unset IFS
    var=`cat ./tests/${mode}/${strarr[0]}.arg`
    final_arg=$arg" "$var
    if ! [ "$mode" = "udp" ]; then
        timeout 1 ./ipkcpc $final_arg < $file > output.txt
    else
        timeout 0.5s sleep 0.5s ./ipkcpc $final_arg < $file > output.txt &
        PID=$!
        kill -SIGINT $PID
    fi
    exit_code=$?
    expected_exit_code=`cat ./tests/${mode}/${strarr[0]}.errcode`
    echo "[TEST] \"${strarr[0]}\" [ARG] \"${final_arg}\""
    if test -f output.txt; then
        var=`diff ./output.txt ./tests/${mode}/${strarr[0]}.out`
        if ! [ "$var" = "" ]; then
            echo "[OUT] Expected output:"
            cat ./tests/${mode}/${strarr[0]}.out
            echo "[OUT] Got output:"
            cat output.txt
            test_not_ok=$((test_not_ok+1))
            continue
        fi
    else
        if [ $exit_code != $expected_exit_code ]; then
            echo "[ERR] File \"output.txt\" was not found"
            test_not_ok=$((test_not_ok+1))
            continue
        fi
    fi
    if [ $exit_code != $expected_exit_code ]
    then
        echo -e "[ERR] Exit code $exit_code, expected: $expected_exit_code"
        test_not_ok=$((test_not_ok+1))
        continue
    fi
    test_ok=$((test_ok+1))
done

echo "[END] $test_ok tests passed, $test_not_ok failed"

tests/create-test.sh
######################################
# 
# Modul:  create-test.sh
#
# Author: Michal Zapletal
# Email:  xzaple41@stud.fit.vutbr.cz
#
######################################

#!/bin/bash

echo -n "Your script name:"
read name
echo -n "Destination [udp/tcp]:"
read mode
touch ./$mode/$name.in
touch ./$mode/$name.out
touch ./$mode/$name.arg
touch ./$mode/$name.errcode
if [ $mode = "udp" ]; then
    echo -n "-m udp" > ./$mode/$name.arg
fi

tests/remove-test.sh
######################################
# 
# Modul:  remove-test.sh
#
# Author: Michal Zapletal
# Email:  xzaple41@stud.fit.vutbr.cz
#
######################################

#!/bin/bash

echo -n "Name of the test you want to delete:"
read name
echo -n "From folder TCP or UDP ? [tcp/udp]:"
read mode
echo -n "Do you really want to delete test \"$name\"? [y/n]"
read opt
if [ "$opt" = "y" ] || [ "$opt" = "Y" ]; then
    file=../tests/$mode/$name.in
    if [[ -f "$file" ]]; then
        rm $file
    fi
    file=../tests/$mode/$name.arg
    if [[ -f "$file" ]]; then
        rm $file
    fi
    file=../tests/$mode/$name.out
    if [[ -f "$file" ]]; then
        rm $file
    fi
    file=../tests/$mode/$name.errcode
    if [[ -f "$file" ]]; then
        rm $file
    fi
fi

ipkcpc.c
////////////////////////////////////////////////////////////////////////////
//
// Modul: ipkcpc.c
//
// Autor: Michal Zapletal
// Email: xzaple41@stud.fit.vutbr.cz
//
// This modul creates terminal app for remote calculator
// Copyright (C) 2023, Zapletal Michal
//
// Source codes inspired from:
// - https://git.fit.vutbr.cz/NESFIT/IPK-Projekty/src/branch/master/Stubs
// - https://www.educative.io/answers/how-to-implement-tcp-sockets-in-c
// - https://www.educative.io/answers/how-to-implement-udp-sockets-in-c
// - And from pdf used in presentations in class IPK
//
///////////////////////////////////////////////////////////////////////////
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <signal.h>
#include <stdbool.h>
#include <unistd.h>
#include "ipkcpc.h"

#ifdef _WIN32
#include <winsock2.h>
#include <ws2tcpip.h>
#include <winsock.h>
#include <conio.h>
#endif // _WIN32

#ifdef __unix__
#include <sys/socket.h>
#include <sys/types.h>
#include <sys/time.h>
#include <arpa/inet.h>
#include <netdb.h>
#endif // unix

//#define DEBUG

// Structure and socket has to be global, because of exit
// if SIGINT is called
ConnectInfo_t * info;

#ifdef _WIN32
SOCKET client_socket;
#endif // _WIN32
#ifdef __unix__
int client_socket;
#endif // unix


/**
 * @brief Function that decodes parametrs from command line, puts them into structure
 *
 * @param argv Arguments from command line
 * @return ConnectInfo_t* Connection structure
 */
void decode_params(int argc, char *argv[]){
    info = malloc(sizeof(ConnectInfo_t));
    if(info == NULL){
        #ifdef DEBUG
        printf("[ERROR] Malloc failure\n");
        #endif
        exit(INTERNAL_ERROR);
    }
    info->IPv4 = NULL;
    info->port = NULL;
    info->connectmode = NONE_MODE;

    if(argc > 1){
        if(strcmp(argv[1],"--help") == 0){
            printf(
            "ipkcpc [options] [details] ...\n\n"
            "OPTIONS\n"
            "\t-h [IP]\n"
            "\t\tSpecify IP address of server to communicate to\n"
            "\t\t[%s by default]\n"
            "\t-m [mode]\n"
            "\t\tSpecify mode of connection. TCP or UDP\n"
            "\t\t[TCP by default]\n"
            "\t-p [port]\n"
            "\t\tSpecify port for connection.\n"
            "\t\t[%s by default]\n\n"
            "DETAILS\n"
            "\tIP\n"
            "\t\tIPv4 address in form \"x.x.x.x\"\n"
            "\t\twhere x is in range <0,255>\n"
            "\tmode\n"
            "\t\tMode of connection. TCP or UDP\n"
            "\tport\n"
            "\t\tPort where it connects to. Number in range <0,65535>\n\n"
            "Author:  Michal Zapletal\n"
            "Contact: xzaple41@stud.fit.vutbr.cz\n",DEFAULT_IPv4,DEFAULT_PORT
            );
            exit(ALL_OK);
        }

        for(int i = 1; i < argc ; ){
            if(strcmp(argv[i],"-h") == 0){
                decode_ip(argv[i+1]);
            }else if(strcmp(argv[i],"-m") == 0){
                decode_mode(argv[i+1]);
            }else if(strcmp(argv[i],"-p") == 0){
                decode_port(argv[i+1]);
            }else{
                #ifdef DEBUG
                printf("[ERROR] Unknown settings\"%s\"\n",argv[i]);
                #endif
                exit(PARAM_ERROR);
            }
            i = i+2;
        }
    }

    //Setting default values if not declared while executing
    if(info->connectmode == NONE_MODE){
        info->connectmode = TCP;
    }
    if(info->IPv4 == NULL){
        info->IPv4 = calloc(sizeof(char),strlen(DEFAULT_IPv4)+1);
        memcpy(info->IPv4,DEFAULT_IPv4,strlen(DEFAULT_IPv4)+1);
    }
    if(info->port == NULL){
        info->port = calloc(sizeof(char),strlen("2023")+1);
        memcpy(info->port,"2023",strlen("2023")+1);
    }
}

/**
 * @brief Function that decodes port, saves port into structure
 *
 * @param port String with port
 */
void decode_port(char * port){
    if(info->port != NULL){
        #ifdef DEBUG
        printf("[ERROR] Double declaration of port\n");
        #endif
        exit(PARAM_ERROR);
    }
    unsigned length = strlen(port);
    if(length > MAX_PORT_CHAR_LENGTH){
        #ifdef DEBUG
        printf("[ERROR] Port contains more than 5 numbers\n");
        #endif
        exit(PARAM_ERROR);
    }
    for(unsigned i = 0; i < length; i++){
        if(!isdigit(port[i])){
            #ifdef DEBUG
            printf("[ERROR] Port must contain only numbers\n");
            #endif
            exit(PARAM_ERROR);
        }
    }
    int portnumber = atoi(port);
    if(portnumber > MAX_PORT_NUM){
        #ifdef DEBUG
        printf("[ERROR] Port number is out of range\n");
        #endif
        exit(PARAM_ERROR);
    }
    info->port = calloc(length+1,sizeof(char));
    if(info->port == NULL){
        #ifdef DEBUG
        printf("[ERROR] Malloc failure\n");
        #endif
        exit(INTERNAL_ERROR);
    }
    memcpy(info->port,port,length);
}

/**
 * @brief Function that decodes mode, saves mode into structure
 *
 * @param mode String with mode
 */
void decode_mode(char * mode){
    if(info->connectmode != NONE_MODE){
        #ifdef DEBUG
        printf("[ERROR] Double declaration of mode\n");
        #endif
        exit(PARAM_ERROR);
    }
    unsigned length = strlen(mode);
    if(length == MODE_CHAR_LENGTH){
        for(unsigned i = 0; i < length; i++){
            if((mode[i] >= 'a' && mode[i] <= 'z') || (mode[i] >= 'A' && mode[i] <= 'Z')){
                mode[i] = toupper(mode[i]);
                continue;
            }else{
                #ifdef DEBUG
                printf("[ERROR] Mode do not contain only characters\n");
                #endif
                exit(PARAM_ERROR);
            }
        }
        if(strcmp(mode,"UDP") == 0){
            info->connectmode = UDP;
        }else if(strcmp(mode,"TCP") == 0){
            info->connectmode = TCP;
        }else{
            #ifdef DEBUG
            printf("[ERROR] Mode has to be UDP or TCP\n");
            #endif
            exit(PARAM_ERROR);
        }
    }else{
        #ifdef DEBUG
        printf("[ERROR] Mode has to be UDP or TCP\n");
        #endif
        exit(PARAM_ERROR);
    }
}

/** Function that decodes ip address.
 *
 * Function check if ip is in right form and puts it into struct
 *
 * @param ip String with ip inside
 */
void decode_ip(char *ip){
    if(info->IPv4 != NULL){
        #ifdef DEBUG
        printf("[ERROR] Double declaration of IPv4\n");
        #endif
        exit(PARAM_ERROR);
    }
    unsigned length = strlen(ip);
    unsigned i = 0;
    unsigned max_range = MAX_IP_COMPONENT_LENGTH; // Boundaries
    unsigned dots_counter = 0;
    for(unsigned j = 0; j < MAX_IP_NUM_OF_COMPONENT; j++){
        if(j != dots_counter){
            #ifdef DEBUG
            printf("[ERROR] IPv4 dont have all elements\n");
            #endif
            exit(PARAM_ERROR);
        }
        for( ; i < length; i++){
            if(!isdigit(ip[i])){
                if(ip[i] == '.'){
                    i++;
                    max_range = i + MAX_IP_COMPONENT_LENGTH;
                    dots_counter++;
                    break;
                }else{
                    #ifdef DEBUG
                    printf("[ERROR] IP address is not created from digits and dots\n");
                    #endif
                    exit(PARAM_ERROR);
                }
            }


            if(i == max_range){
                #ifdef DEBUG
                printf("[ERROR] Wrong IP address\n");
                #endif
                exit(PARAM_ERROR);
            }
        }
    }
    info->IPv4 = calloc(length+1,sizeof(char));
    if(info->IPv4 == NULL){
        #ifdef DEBUG
        printf("[ERROR] Malloc failure\n");
        #endif
        exit(INTERNAL_ERROR);
    }
    memcpy(info->IPv4,ip,length);
    check_ip(info->IPv4);
}

/**
 * @brief Function that check boundaries of IP, if all numbers
 * are in range <0,255>
 *
 * @param ip String with ip address
 */
void check_ip(char * ip){
    int length = 0;
    char * to_int = malloc((length+1)*sizeof(char));
    if(to_int == NULL){
        #ifdef DEBUG
        printf("[ERROR] Malloc failure\n");
        #endif
        exit(INTERNAL_ERROR);
    }
    for(unsigned i = 0; i <= strlen(ip); i++){
        if(isdigit(ip[i])){
            length++;
            char * new_to_int = realloc(to_int,(length+1)*sizeof(char));
            if(new_to_int == NULL){
                #ifdef DEBUG
                printf("[ERROR] Realloc failure\n");
                #endif
                exit(INTERNAL_ERROR);
            }
            to_int = new_to_int;
            to_int[length-1] = ip[i];
            to_int[length] = '\0';
        }else{
            int var = atoi(to_int);
            if(var > MAX_IP_COMPONENT_VALUE){
                #ifdef DEBUG
                printf("[ERROR] Value of IP out of range <0,255>\n");
                #endif
                exit(PARAM_ERROR);
            }
            length = 0;
        }
    }
    free(to_int);
}

/**
 * @brief Connection struct dectructor
 */
void struct_dtor(){
    if(info != NULL){
        if(info->IPv4 != NULL){
            free(info->IPv4);
        }
        if(info->port != NULL){
            free(info->port);
        }
        free(info);
    }
}

/**
 * @brief Create a socket for TCP connection
 *
 * @return Server socket
 */
void create_socket_tcp(){
    int family = AF_INET;
    int type = SOCK_STREAM;
    #ifdef _WIN32
    WSADATA wsa;
    char c = 0;
    if (WSAStartup(MAKEWORD(2,2), &wsa) != 0)
    {
        #ifdef DEBUG
        printf("[ERROR] wsa\n");
        #endif
        exit(WSA_FAILURE);
    }
    client_socket = socket(family, type, IPPROTO_TCP);
    (void)c;
    #endif
    #ifdef __unix__
    client_socket = socket(family, type, 0);
    #endif
    if(client_socket <= 0){
        #ifdef DEBUG
        printf("[ERROR] Socket failure\n");
        #endif
        exit(SOCKET_FAILURE);
    }

    struct sockaddr_in server_addr;

    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(atoi(info->port));
    server_addr.sin_addr.s_addr = inet_addr(info->IPv4);

    if(connect(client_socket, (struct sockaddr*)&server_addr, sizeof(server_addr)) < 0){
        #ifdef DEBUG
        printf("[ERROR] Unable to connect\n");
        #endif
        exit(CONNECTION_FAILURE);
    }
}

/**
 * @brief Create connection with declared server, on declared port, TCP.
 * In infinite loop sending messages to server until connection is closed
 */
void send_message_tcp(){
    char client_message[MAX_MESSAGE_LENGTH];
    char server_message[MAX_MESSAGE_LENGTH];
    char message_to_print[MAX_MESSAGE_LENGTH];

    create_socket_tcp();

    while(true){
        memset(client_message,'\0',MAX_MESSAGE_LENGTH);
        memset(server_message,'\0',MAX_MESSAGE_LENGTH);
        memset(message_to_print,'\0',MAX_MESSAGE_LENGTH);
        for(int i = 0; ; i++){
            int c = fgetc(stdin);
            client_message[i] = c;
            if(i == MAX_MESSAGE_LENGTH-1){
                #ifdef DEBUG
                printf("[ERROR] Input message is too long!\n");
                exit(TOO_LONG_MESSAGE);
                #endif
            }
            if(c == '\n'){
                break;
            }
        }
        int bytes_tx = send(client_socket,client_message,strlen(client_message), 0);
        if(bytes_tx < 0){
            #ifdef DEBUG
            printf("[ERROR] sendto failure\n");
            #endif
            exit(SENDTO_FAILURE);
        }
        // TCP will recieve until LF is read
        while(true){
            int bytes_rx = recv(client_socket,server_message,MAX_MESSAGE_LENGTH,0);
            if(bytes_rx < 0){
                #ifdef DEBUG
                printf("[ERROR] Recieve Failure\n");
                #endif
                exit(RECIEVE_FAILURE);
            }
            strcat(message_to_print,server_message);
            memset(server_message,'\0',MAX_MESSAGE_LENGTH);
            if(message_to_print[strlen(message_to_print)-1] == '\n'){
                break;
            }
        }
        if(strcmp(message_to_print,SERVER_ENDING_MESSAGE) == 0){
            printf("%s",message_to_print);
            break;
        }else{
            printf("%s",message_to_print);
        }
    }

    close(client_socket);
    #ifdef _WIN32
    WSACleanup();
    #endif
}
/**
 * @brief Create a socket for udp communication
 * 
 */
void create_socket_udp(){
    #ifdef _WIN32
    WSADATA wsa;
    if (WSAStartup(MAKEWORD(2,2), &wsa) != 0)
    {
        #ifdef DEBUG
        printf("[ERROR] wsa\n");
        #endif
        exit(WSA_FAILURE);
    }
    #endif

    client_socket = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
    if(client_socket <= 0){
        #ifdef DEBUG
        printf("[ERROR] Socket failure\n");
        #endif
        exit(SOCKET_FAILURE);
    }

    struct timeval timeout;
    timeout.tv_sec = UDP_TIMEOUT;
    timeout.tv_usec = 0;

    #ifdef __unix__
    int sock_opt = setsockopt(client_socket,SOL_SOCKET, SO_RCVTIMEO,&timeout,sizeof(timeout));
    #endif
    #ifdef _WIN32
    int sock_opt = setsockopt(client_socket,SOL_SOCKET, SO_RCVTIMEO,(const char *)&timeout,sizeof(timeout));
    #endif
    if(sock_opt < 0){
        #ifdef DEBUG
        printf("[ERROR] Set socket option failure\n");
        #endif
        exit(SOCKET_FAILURE);
    }
}
/**
 * @brief Function that send socket with UDP mode
 * 
 */
void send_message_udp(){
    create_socket_udp();

    struct hostent * server = gethostbyname(info->IPv4);
    if(server == NULL){
        #ifdef DEBUG
        printf("[ERROR] Server IP error\n");
        #endif
        exit(SENDTO_FAILURE);
    }

    struct sockaddr_in server_address;
    memset(&server_address,0,sizeof(server_address));
    server_address.sin_family = AF_INET;
    server_address.sin_port = htons(atoi(info->port));
    server_address.sin_addr.s_addr = inet_addr(info->IPv4);
    socklen_t addr_size = sizeof(server_address);

    char client_message[MAX_MESSAGE_LENGTH];
    char server_message[MAX_MESSAGE_LENGTH];

    while(true){
        memset(client_message,'\0',MAX_MESSAGE_LENGTH);
        memset(server_message,'\0',MAX_MESSAGE_LENGTH);
        for(int i = UDP_SKIP_BYTES_SEND; ; i++){
            int c = fgetc(stdin);
            if(i == MAX_MESSAGE_LENGTH-1){
                #ifdef DEBUG
                printf("[ERROR] Input message is too long!\n");
                exit(TOO_LONG_MESSAGE);
                #endif
            }
            if(c == '\n'){
                break;
            }
            client_message[i] = c;
        }
        client_message[UDP_OPCODE_POS] = UDP_OPCODE_REQUEST;
        client_message[1] = strlen(&(client_message[UDP_SKIP_BYTES_SEND]));
        int bytes_tx = sendto(client_socket,client_message,
                            strlen(&(client_message[UDP_SKIP_BYTES_SEND]))+UDP_SKIP_BYTES_SEND,
                            0,(struct sockaddr *)&server_address,addr_size);
        if(bytes_tx < 0){
            #ifdef DEBUG
            printf("[ERROR] sendto failure\n");
            #endif
            exit(SENDTO_FAILURE);
        }

        int bytes_rx = recvfrom(client_socket,server_message,MAX_MESSAGE_LENGTH,0,
                            (struct sockaddr *)&server_address,&addr_size);
        if(bytes_rx < 0){
            #ifdef DEBUG
            printf("[ERROR] Recieve Failure\n");
            #endif
            exit(RECIEVE_FAILURE);
        }
        if(server_message[UDP_OPCODE_POS] == UDP_OPCODE_REQUEST){
            printf("ERR: Expected message from server with opcode \"RESPONSE\"\n");
        }else if(server_message[UDP_OPCODE_POS] == UDP_OPCODE_RESPONSE){
            if(server_message[UDP_STATUS_CODE_POS] == UDP_STATUS_CODE_OK){
                printf("OK:%s\n",&(server_message[UDP_SKIP_BYTES_RECV]));
            }else if(server_message[UDP_STATUS_CODE_POS] == UDP_STATUS_CODE_ERR){
                printf("ERR:%s\n",&(server_message[UDP_SKIP_BYTES_RECV]));
            }else{
                #ifdef DEBUG
                printf("[ERROR] Wrong OPCODE\n");
                #endif
                exit(WRONG_STATUS_CODE);
            }
        }else{
            #ifdef DEBUG
            printf("[ERROR] Wrong OPCODE\n");
            #endif
            exit(WRONG_OPCODE);
        }
    }
}

/**
 * @brief Function for handling interruption
 * 
 * @param _ EMPTY
 */
static void sigint_handler(int _){
    signal(SIGINT, sigint_handler);
    if(info->connectmode == TCP){
        char client_message[MAX_MESSAGE_LENGTH] = SERVER_ENDING_MESSAGE;
        char server_message[MAX_MESSAGE_LENGTH];
        char message_to_print[MAX_MESSAGE_LENGTH];
        memset(server_message,'\0',MAX_MESSAGE_LENGTH);
        memset(message_to_print,'\0',MAX_MESSAGE_LENGTH);
        int bytes_tx = send(client_socket,client_message,strlen(client_message), 0);
        if(bytes_tx < 0){
            #ifdef DEBUG
            printf("[ERROR] sendto failure\n");
            #endif
            exit(SENDTO_FAILURE);
        }
        while(true){
            int bytes_rx = recv(client_socket,server_message,MAX_MESSAGE_LENGTH,0);
            if(bytes_rx < 0){
                #ifdef DEBUG
                printf("[ERROR] Recieve Failure\n");
                #endif
                exit(RECIEVE_FAILURE);
            }
            strcat(message_to_print,server_message);
            memset(server_message,'\0',MAX_MESSAGE_LENGTH);
            if(message_to_print[strlen(message_to_print)-1] == '\n'){
                printf("\n%s",message_to_print);
                break;
            }
        }

        close(client_socket);
    }
    #ifdef _WIN32
    WSACleanup();
    #endif
    struct_dtor();
    (void)_;
    exit(ALL_OK);
}

int main(int argc, char *argv[]){
    signal(SIGINT, sigint_handler);

    decode_params(argc,argv);

    #ifdef DEBUG
    printf("IP:   %s\n",info->IPv4);
    printf("PORT: %s\n",info->port);
    if(info->connectmode == UDP){
        printf("MODE: UDP\n");
    }else if(info->connectmode == TCP){
        printf("MODE: TCP\n");
    }else{
        printf("MODE: none\n");
    }
    #endif

    if(info->connectmode == TCP){
        send_message_tcp(info);
    }else{
        send_message_udp(info);
    }

    struct_dtor();
    exit(ALL_OK);
}


ipkcpc.h
/////////////////////////////////////////////////////////
//
// Modul: ipkcpc.h
//
// Autor: Michal Zapletal
// Email: xzaple41@stud.fit.vutbr.cz
//
// Library for ipkcpc.c file
// Copyright (C) 2023, Zapletal Michal
//
/////////////////////////////////////////////////////////

#ifndef _IPK_PROJ1_LIB_
#define _IPK_PROJ1_LIB_

// Port number (from his definition, 16-bit number) has to be in range <0,65535>
#define MAX_PORT_NUM 65535

// Randomly choosed maximal length of message that can be sent(in bytes)
#define MAX_MESSAGE_LENGTH 256

// Default port to connect to, same value as reference server has as default
#define DEFAULT_PORT "2023"

// Default IPv4 to connect to, same value as reference server has as default
#define DEFAULT_IPv4 "0.0.0.0"

// Definiton of maximal length(chars) of port
#define MAX_PORT_CHAR_LENGTH 5

// TCP or UDP are created by 3 characters
#define MODE_CHAR_LENGTH 3

/** Components of IP address are created of numbers, IP address is X.X.X.X,
 * where X is in range <0,255>, where 255 is length of 3
 */
#define MAX_IP_COMPONENT_LENGTH 3

// IPv4 address is in shape X.X.X.X, wher X is four times
#define MAX_IP_NUM_OF_COMPONENT 4

// Components of IP address are in range <0,255>
#define MAX_IP_COMPONENT_VALUE 255

// TCP connection is terminated by server by sending this message
#define SERVER_ENDING_MESSAGE "BYE\n"

// From its definition (described in documentation) we skip first 3 bytes from recieved message
#define UDP_SKIP_BYTES_RECV 3

// From its definition (described in documentation) we skip first 2 bytes from sending message
#define UDP_SKIP_BYTES_SEND 2

// Definition of opcodes
#define UDP_OPCODE_REQUEST 0
#define UDP_OPCODE_RESPONSE 1

// Definiton of status codes
#define UDP_STATUS_CODE_OK 0
#define UDP_STATUS_CODE_ERR 1

// Positions of STATUS CODES and OPCODES
#define UDP_OPCODE_POS 0
#define UDP_STATUS_CODE_POS 1

// Timeout for udp connection, in seconds
#define UDP_TIMEOUT 5

// Enum for connection types
typedef enum{
    UDP,
    TCP,
    NONE_MODE
}MODE;

// Error numbers
typedef enum{
    ALL_OK,
    PARAM_ERROR,
    SOCKET_FAILURE,
    SENDTO_FAILURE,
    CONNECTION_FAILURE,
    RECIEVE_FAILURE,
    WRONG_OPCODE,
    WRONG_STATUS_CODE,
    WSA_FAILURE,
    TOO_LONG_MESSAGE,
    INTERNAL_ERROR = 99
} EXIT_STATUSES;

/**
 * @brief Structure that saves communication details
 * 
 * @param IPv4 IPv4 address
 * @param port port number
 * @param connectmode TCP or UDP mode
 */
typedef struct ConnectionInfo ConnectInfo_t;
struct ConnectionInfo{
    char *IPv4;
    char *port;
    MODE connectmode;
};

void decode_params(int argc,char *argv[]);
void decode_port(char * port);
void decode_mode(char * mode);
void decode_ip(char *ip);
void check_ip(char * ip);
void struct_dtor();
void create_socket_tcp();
void send_message_tcp();
void create_socket_udp();
void send_message_udp();
static void sigint_handler(int _);

#endif